File: Jeval/Compiler.java
package Jeval;

import java.util.ArrayList;

public class Compiler {

    public enum ItemType {
        SYM, CST, VAR, EXP, FNC
    }

    public static class Item {
        public String name;
        public String string;
        public int register = -1;
        public double value;
        public ItemType type;

        public String toString() {
            if (this.type == ItemType.CST) {
                return "[CST]" + name + "<" + value + ">";
            }
            return "[" + type + "]" + name;
        }
    }

    public static class Const extends Item {

        Const(String string) {
            this.type = ItemType.CST;
            this.value = Double.valueOf(string);
            this.string = string;
            this.name = string;
        }

        Const(double value) {
        }
    }

    public static class Symbol extends Item {
        Symbol(String string) {
            this.type = ItemType.SYM;
            this.string = string;
            this.name = string;
        }

        Symbol(char string) {
            this.type = ItemType.SYM;
            this.string = String.valueOf(string);
            this.name = this.string;
        }
    }

    public static class Variable extends Item {
        Variable(String string) {
            this.type = ItemType.VAR;
            this.string = string;
            this.name = string;

            for (String func : functionList) {
                if (func.equals(string)) {
                    type = ItemType.FNC;
                }
            }
        }
    }

    public static class Expression extends Item {
        Expression(String string) {
            this.type = ItemType.EXP;
            this.string = string;
            this.name = string;
        }
    }

    public static char[] symbolList = "+-*/^".toCharArray();
    public static char[] numberList = "0123456789.".toCharArray();
    public static char[] alphabate = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".toCharArray();
    public static String[] functionList = new String[] {
            "sin", "cos", "tan",
            "log", "ln", "sqrt"
    };

    public boolean mandatoryConstants = false;
    public int maxRegisterNumber = 0;
    public int autoInitRegister = 16;
    public ArrayList<Command> commandList = new ArrayList<>();
    public ArrayList<Boolean> registerList = new ArrayList<>();

    public int getFreeRegister() {
        for (int i = 0; i < registerList.size(); i++) {
            if (!registerList.get(i)) {
                registerList.set(i, true);
                return i;
            }
        }
        registerList.add(true);
        maxRegisterNumber = Math.max(maxRegisterNumber, registerList.size());
        return registerList.size() - 1;
    }

    public void freeRegister(int index) {
        registerList.set(index, false);
    }

    public void o(Command cmd) {
        commandList.add(cmd);
    }

    public static boolean isNumber(char value) {
        for (char key : numberList) {
            if (key == value)
                return true;
        }
        return false;
    }

    public static boolean isAlphabet(char value) {
        for (char key : alphabate) {
            if (key == value)
                return true;
        }
        return false;
    }

    public static boolean isSymbol(char value) {
        for (char key : symbolList) {
            if (key == value)
                return true;
        }
        return false;
    }

    public static void printItemList(ArrayList<Item> ItemList) {
        System.out.print("{");
        for (Item item : ItemList) {
            System.out.print(item.toString());
            System.out.print(", ");
        }
        System.out.println("}");
    }

    public static double ln(double x) {
        return Math.log(x) / Math.log(Math.E);
    }

    public static Command[] toCommandArray(ArrayList<Command> arrayList) {
        Command[] array = new Command[arrayList.size()];
        for (int i = 1; i < arrayList.size(); i++) {
            array[i] = arrayList.get(i);
        }
        return array;
    }

    public static ArrayList<Item> split(String expr) {
        int startPoint = 0;
        int braStartPoint = 0;
        int inBracket = 0;
        boolean inNumber = false;
        boolean inAlpha = false;
        ArrayList<Item> results = new ArrayList<>();
        char key;
        int i = 0;

        expr = expr + " ";
        if (expr.charAt(0) == '-')
            expr = "0" + expr;

        while (i < expr.length()) {
            key = expr.charAt(i);
            if (key == ')') {
                inBracket--;
                if (inBracket == 0) {
                    Item item = new Compiler.Expression(expr.substring(braStartPoint + 1, i));
                    results.add(item);
                }
            } else if (key == '(') {
                inBracket++;
                if (inBracket == 1)
                    braStartPoint = i;
            } else if (inBracket != 0) {
                i++;
                continue;
            }

            if (inNumber && !isNumber(key)) {
                inNumber = false;
                Const item = new Const(expr.substring(startPoint, i));
                results.add(item);
            } else if (!inNumber && !inAlpha && isNumber(key)) {
                inNumber = true;
                startPoint = i;
            }
            if (inAlpha && !(isAlphabet(key) || isNumber(key))) {
                inAlpha = false;
                Item item = new Variable(expr.substring(startPoint, i));
                results.add(item);
            } else if (!inAlpha && isAlphabet(key)) {
                inAlpha = true;
                startPoint = i;
            }
            if (isSymbol(key)) {
                Item item = new Symbol(key);
                results.add(item);
            }
            i++;
        }
        return results;
    }

    public static ArrayList<Item> addMulSymbol(ArrayList<Item> itemList) {
        int i = 0;
        Item thisObj;
        Item lastObj;
        while (i < itemList.size() - 1) {
            i++;
            thisObj = itemList.get(i);
            lastObj = itemList.get(i - 1);
            if (thisObj.type == ItemType.VAR || thisObj.type == ItemType.EXP || thisObj.type == ItemType.FNC) {
                if (lastObj.type == ItemType.VAR || lastObj.type == ItemType.CST || lastObj.type == ItemType.EXP) {
                    itemList.add(i, new Symbol("*"));
                    i++;
                }
            }
        }
        return itemList;
    }

    private ArrayList<Item> calculate(ArrayList<Item> itemList) throws Exception {

        // process all the vars
        for (Item item : itemList) {
            if (item.type == ItemType.VAR) {
                int register = getFreeRegister();
                item.register = register;
                o(new Command.InitVar(item.name, register));
            }
        }

        int i = 0;
        while (i < itemList.size() - 1) {
            Item item = itemList.get(i);
            Item next = itemList.get(i + 1);
            if (item.type == ItemType.FNC && next.type == ItemType.CST) {
                Item itemfill;
                if (item.name.equals("sin")) {
                    itemfill = new Const(Math.sin(next.value));
                } else if (item.name.equals("cos")) {
                    itemfill = new Const(Math.cos(next.value));
                } else if (item.name.equals("tan")) {
                    itemfill = new Const(Math.tan(next.value));
                } else if (item.name.equals("log")) {
                    itemfill = new Const(Math.log(next.value));
                } else if (item.name.equals("ln")) {
                    itemfill = new Const(ln(next.value));
                } else if (item.name.equals("sqrt")) {
                    itemfill = new Const(Math.sqrt(next.value));
                } else {
                    throw new Exception("Unknown function name: " + item.name);
                }

                itemList.set(i, itemfill);
                itemList.remove(i + 1);
                i++;
            } else if (item.type == ItemType.FNC) {
                if (item.name.equals("sin")) {
                    o(new Command.Sin(next.register, next.register));
                } else if (item.name.equals("cos")) {
                    o(new Command.Cos(next.register, next.register));
                } else if (item.name.equals("tan")) {
                    o(new Command.Tan(next.register, next.register));
                } else if (item.name.equals("log")) {
                    o(new Command.Log(next.register, next.register));
                } else if (item.name.equals("ln")) {
                    o(new Command.Ln(next.register, next.register));
                } else if (item.name.equals("sqrt")) {
                    o(new Command.Sqrt(next.register, next.register));
                } else {
                    throw new Exception("Unknown function name: " + item.name);
                }

                Item itemfill = new Variable("temp");
                itemfill.register = next.register;
                itemList.set(i, itemfill);
                itemList.remove(i + 1);
                i++;
            }
            i++;
        }

        i = 0;
        while (i < itemList.size()) {
            Item item = itemList.get(i);
            int symbolType;
            if (item.type == ItemType.SYM) {

                if (i == 0 || i == itemList.size() - 1)
                    throw new Exception("Binocular operators cannot be at the beginning/end");
                if (item.name.equals("*")) {
                    symbolType = 1;
                } else if (item.name.equals("/")) {
                    symbolType = 2;
                } else {
                    i++;
                    continue;
                }

                Item lastItem = itemList.get(i - 1);
                Item nextItem = itemList.get(i + 1);
                Item itemfill = null;
                if (lastItem.register == -1 && nextItem.register == -1) {
                    if (symbolType == 1) {
                        itemfill = new Const(lastItem.value * nextItem.value);
                    } else {
                        itemfill = new Const(lastItem.value / nextItem.value);
                    }
                } else if (lastItem.register != -1 && nextItem.register == -1) {
                    int ro = lastItem.register;
                    if (symbolType == 1) {
                        o(new Command.Time(nextItem.value, ro));
                    } else {
                        o(new Command.Time(1 / nextItem.value, ro));
                    }
                    itemfill = new Variable("temp");
                    itemfill.register = ro;
                } else if (lastItem.register == -1 && nextItem.register != -1) {
                    int ro = nextItem.register;
                    if (symbolType == 1) {
                        o(new Command.Time(lastItem.value, ro));
                    } else {
                        o(new Command.Time(1 / lastItem.value, ro));
                    }
                    itemfill = new Variable("temp");
                    itemfill.register = ro;
                } else {
                    int r1 = lastItem.register;
                    int r2 = nextItem.register;
                    if (symbolType == 1) {
                        o(new Command.Mul(r1, r2, r2));
                    } else {
                        o(new Command.Div(r1, r2, r2));
                    }
                    freeRegister(r1);
                    itemfill = new Variable("temp");
                    itemfill.register = r2;
                }
                itemList.set(i, itemfill);
                itemList.remove(i - 1);
                itemList.remove(i);
            } else {
                i++;
            }
        }

        i = 0;
        while (i < itemList.size()) {
            Item item = itemList.get(i);
            int symbolType;
            if (item.type == ItemType.SYM) {

                if (i == 0 || i == itemList.size() - 1)
                    throw new Exception("Binocular operators cannot be at the beginning/end");
                if (item.name.equals("+")) {
                    symbolType = 1;
                } else if (item.name.equals("-")) {
                    symbolType = 2;
                } else {
                    i++;
                    continue;
                }

                Item lastItem = itemList.get(i - 1);
                Item nextItem = itemList.get(i + 1);
                Item itemfill = null;
                if (lastItem.register == -1 && nextItem.register == -1) {
                    if (symbolType == 1) {
                        itemfill = new Const(lastItem.value + nextItem.value);
                    } else {
                        itemfill = new Const(lastItem.value - nextItem.value);
                    }
                } else if (lastItem.register != -1 && nextItem.register == -1) {
                    int ro = lastItem.register;
                    if (symbolType == 1) {
                        o(new Command.Accumulate(nextItem.value, ro));
                    } else {
                        o(new Command.Accumulate(0 - nextItem.value, ro));
                    }
                    itemfill = new Variable("temp");
                    itemfill.register = ro;
                } else if (lastItem.register == -1 && nextItem.register != -1) {
                    int ro = nextItem.register;
                    if (symbolType == 1) {
                        o(new Command.Accumulate(lastItem.value, ro));
                    } else {
                        o(new Command.Accumulate(1 / lastItem.value, ro));
                    }
                    itemfill = new Variable("temp");
                    itemfill.register = ro;
                } else {
                    int r1 = lastItem.register;
                    int r2 = nextItem.register;
                    if (symbolType == 1) {
                        o(new Command.Add(r1, r2, r2));
                    } else {
                        o(new Command.Min(r1, r2, r2));
                    }
                    freeRegister(r1);
                    itemfill = new Variable("temp");
                    itemfill.register = r2;
                }
                itemList.set(i, itemfill);
                itemList.remove(i - 1);
                itemList.remove(i);
            } else {
                i++;
            }
        }

        return itemList;
    }

    private Item autoCompileLauncher(String expr) throws Exception {
        ArrayList<Item> itemList = split(expr);
        itemList = addMulSymbol(itemList);

        for (int i = 0; i < itemList.size(); i++) {
            Item item = itemList.get(i);
            if (item.type == ItemType.EXP) {
                Item j = autoCompileLauncher(item.string);
                if (j.type == ItemType.CST) {
                    itemList.set(i, new Const(j.value));
                } else {
                    item.register = j.register;
                }
            }
        }

        itemList = calculate(itemList);
        if (!(itemList.size() == 1)) {
            throw new Exception("Two separate expressions were entered.");
        } else {
            return itemList.get(0);
        }
    }

    public Command[] compiler(String expr) throws Exception {
        commandList = new ArrayList<Command>();
        registerList = new ArrayList<>();
        maxRegisterNumber = 0;

        try {
            Item i = autoCompileLauncher(expr);
            if (i.type == ItemType.CST) {
                o(new Command.ReturnNumber(i.value));
            } else {
                o(new Command.Return(i.register));
            }

            if (maxRegisterNumber > autoInitRegister) {
                Command command = new Command.InitRegister(maxRegisterNumber);
                commandList.add(0, command);
            }
        } catch (Exception e) {
            throw e;
        }

        return commandList.toArray(new Command[commandList.size()]);
    }
}

File: Peval/compiler.py
from enum import Enum
from Peval.runtime import *


class Compiler:

    class ItemType(Enum):
        SYM = "sym"
        CST = "cst"
        VAR = "var"
        EXP = "exp"
        FNC = "fnc"

    class Item:
        register = None
        string = None
        name = None
        type = None

        def __repr__(self):
            return f"[{self.type.name}]{self.name}"

    class Const(Item):
        def __init__(self, string):
            self.type = Compiler.ItemType.CST
            self.value = float(string)
            self.string = string
            self.name = string

    class Symbol(Item):
        def __init__(self, string):
            self.type = Compiler.ItemType.SYM
            self.string = string
            self.name = string

    class Variable(Item):
        def __init__(self, string):
            self.type = Compiler.ItemType.VAR
            self.string = string
            self.name = string

            for i in Compiler.functionList:
                if i == self.string:
                    self.type = Compiler.ItemType.FNC

    class Expression(Item):
        def __init__(self, string):
            self.type = Compiler.ItemType.EXP
            self.string = string
            self.name = string

    symbolList = list("+-*/^")
    numberList = list("0123456789.")
    alphabeta = list("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
    functionList = [
        "sin", "cos", "tan",
        "log", "ln", "sqrt"
    ]

    commandList = []
    mandatoryConstants = False
    maxRegisterNumber = 0
    autoInitRegister = 16
    registerList = [False for i in range(autoInitRegister)]

    def getFreeRegister(self):
        for i in range(len(self.registerList)):
            if not self.registerList[i]:
                self.registerList[i] = True
                return i
        self.registerList.append(True)
        self.maxRegisterNumber = max(self.maxRegisterNumber, len(self.registerList))
        return len(self.registerList) - 1

    def freeRegister(self, index:int):
        self.registerList[index] = False

    def o(self, command):
        self.commandList.append(command)

    def isNumber(self, value:str):
        return value in self.numberList

    def isAlphabet(self, value:str):
        return value in self.alphabeta

    def isSymbol(self, value:str):
        return value in self.symbolList

    def split(self, expr:str):

        startPoint = None
        braStartPoint = None
        inBracket = 0
        inNumber = False
        inAlpha = False
        results = []
        i = 0

        # process negative symbol
        # process end of expr
        expr = expr + " "
        if expr[0] == "-":
            expr = "0"+expr

        while i < len(expr):
            key = expr[i]
            if key==")":
                inBracket -= 1
                if inBracket == 0:
                    item = Compiler.Expression(expr[braStartPoint+1:i])
                    results.append(item)
            elif key=="(":
                inBracket += 1
                if inBracket == 1:
                    braStartPoint = i
            elif inBracket != 0:
                i = i+1
                continue

            if inNumber and not self.isNumber(key):
                inNumber = False
                item = Compiler.Const(expr[startPoint:i])
                results.append(item)
            elif not inNumber and not inAlpha and self.isNumber(key):
                inNumber = True
                startPoint = i
            if inAlpha and not (self.isAlphabet(key) or self.isNumber(key)):
                inAlpha = False
                item = Compiler.Variable(expr[startPoint:i])
                results.append(item)
            elif not inAlpha and self.isAlphabet(key):
                inAlpha = True
                startPoint = i
            if self.isSymbol(key):
                item = Compiler.Symbol(key)
                results.append(item)

            i = i+1
        return results

    def addMulSymbol(self, itemList:list):
        i = 0
        while i < len(itemList)-1:
            i += 1
            thisObj = itemList[i]
            lastObj = itemList[i-1]
            if thisObj.type == Compiler.ItemType.VAR or thisObj.type == Compiler.ItemType.FNC or thisObj.type == Compiler.ItemType.EXP:
                if lastObj.type == Compiler.ItemType.VAR or lastObj.type == Compiler.ItemType.CST or lastObj.type == Compiler.ItemType.EXP:
                    itemList.insert(i, Compiler.Symbol("*"))
                    i += 1
        return itemList

    def calculate(self, itemList:list):
        # process all the vars
        varList = []
        regList = []
        for i in range(len(itemList)):
            item = itemList[i]
            if item.type == Compiler.ItemType.VAR:
                register = self.getFreeRegister()
                item.register = register
                self.o(InitVar(item.name, register))

        i = 0
        while i < len(itemList)-1:
            item = itemList[i]
            nextItem = itemList[i+1]
            if not item.type == Compiler.ItemType.FNC:
                i += 1
                continue

            if item.name == "sin":
                commandClass = Sin
            elif item.name == "cos":
                commandClass = Cos
            elif item.name == "tan":
                commandClass = Tan
            elif item.name == "log":
                commandClass = Log
            elif item.name == "ln":
                commandClass = Ln
            elif item.name == "sqrt":
                commandClass = Sqrt
            else:
                raise NameError("Unsupported function:", item.name)

            if nextItem.register is None:
                value = commandClass(-1, -1).func(nextItem.value)
                item = Compiler.Const(str(value))
            else:
                item = Compiler.Variable("temp")
                item.register = nextItem.register
                self.o(commandClass(nextItem.register, nextItem.register))
            itemList[i] = item
            itemList.pop(i+1)
            i += 1

        i = 0
        while i < len(itemList):
            item = itemList[i]
            if item.type == Compiler.ItemType.SYM:
                if i == 0 or i == len(itemList)-1:
                    raise Exception("Binocular operators cannot be at the beginning/end")
                if item.name == "*":
                    symbolType = 1
                elif item.name == "/":
                    symbolType = 2
                else:
                    i += 1
                    continue

                last_item = itemList[i - 1]
                next_item = itemList[i + 1]
                if last_item.register is None and next_item.register is None:
                    if symbolType == 1:
                        item = Compiler.Const(last_item.value * next_item.value)
                    else:
                        item = Compiler.Const(last_item.value / next_item.value)
                elif last_item.register is not None and next_item.register is None and self.mandatoryConstants:
                    r1 = self.getFreeRegister()
                    r2 = last_item.register
                    self.o(Const(next_item.value, r1))
                    if symbolType == 1:
                        self.o(Mul(r1, r2, r2))
                    else:
                        self.o(Div(r1, r2, r2))
                    self.freeRegister(r1)
                    item = Compiler.Variable("temp")
                    item.register = r2
                elif last_item.register is not None and next_item.register is None:
                    ro = last_item.register
                    if symbolType == 1:
                        self.o(Time(next_item.value, ro))
                    else:
                        self.o(Time(1/next_item.value, ro))
                    item = Compiler.Variable("temp")
                    item.register = ro
                elif last_item.register is None and next_item.register is not None and self.mandatoryConstants:
                    r1 = self.getFreeRegister()
                    r2 = next_item.register
                    self.o(Const(last_item.value, r1))
                    if symbolType == 1:
                        self.o(Mul(r1, r2, r2))
                    else:
                        self.o(Div(r1, r2, r2))
                    self.freeRegister(r1)
                    item = Compiler.Variable("temp")
                    item.register = r2
                elif last_item.register is None and next_item.register is not None:
                    ro = next_item.register
                    if symbolType == 1:
                        self.o(Time(last_item.value, ro))
                    else:
                        self.o(Time(1/last_item.value, ro))
                    item = Compiler.Variable("temp")
                    item.register = ro
                else:
                    r1 = last_item.register
                    r2 = next_item.register
                    if symbolType == 1:
                        self.o(Mul(r1, r2, r2))
                    else:
                        self.o(Div(r1, r2, r2))
                    self.freeRegister(r1)
                    item = Compiler.Variable("temp")
                    item.register = r2
                itemList[i] = item
                itemList.pop(i-1)
                itemList.pop(i)
            else:
                i = i+1



        i = 0
        while i < len(itemList):
            item = itemList[i]
            if item.type == Compiler.ItemType.SYM:
                if i == 0 or i == len(itemList)-1:
                    raise Exception("Binocular operators cannot be at the beginning/end")
                if item.name == "+":
                    symbolType = 1
                elif item.name == "-":
                    symbolType = 2
                else:
                    i += 1
                    continue

                last_item = itemList[i - 1]
                next_item = itemList[i + 1]
                if last_item.register is None and next_item.register is None:
                    if symbolType == 1:
                        item = Compiler.Const(last_item.value + next_item.value)
                    else:
                        item = Compiler.Const(last_item.value - next_item.value)
                elif last_item.register is not None and next_item.register is None and self.mandatoryConstants:
                    r1 = self.getFreeRegister()
                    r2 = last_item.register
                    self.o(Const(next_item.value, r1))
                    if symbolType == 1:
                        self.o(Add(r1, r2, r2))
                    else:
                        self.o(Min(r1, r2, r2))
                    self.freeRegister(r1)
                    item = Compiler.Variable("temp")
                    item.register = r2
                elif last_item.register is not None and next_item.register is None:
                    ro = last_item.register
                    if symbolType == 1:
                        self.o(Accumulate(next_item.value, ro))
                    else:
                        self.o(Accumulate(0-next_item.value, ro))
                    item = Compiler.Variable("temp")
                    item.register = ro
                elif last_item.register is None and next_item.register is not None and self.mandatoryConstants:
                    r1 = self.getFreeRegister()
                    r2 = next_item.register
                    self.o(Const(last_item.value, r1))
                    if symbolType == 1:
                        self.o(Add(r1, r2, r2))
                    else:
                        self.o(Min(r1, r2, r2))
                    self.freeRegister(r1)
                    item = Compiler.Variable("temp")
                    item.register = r2
                elif last_item.register is None and next_item.register is not None:
                    ro = next_item.register
                    if symbolType == 1:
                        self.o(Accumulate(last_item.value, ro))
                    else:
                        self.o(Accumulate(0-last_item.value, ro))
                    item = Compiler.Variable("temp")
                    item.register = ro
                else:
                    r1 = last_item.register
                    r2 = next_item.register
                    if symbolType == 1:
                        self.o(Add(r1, r2, r2))
                    else:
                        self.o(Min(r1, r2, r2))
                    self.freeRegister(r1)
                    item = Compiler.Variable("temp")
                    item.register = r2
                itemList[i] = item
                itemList.pop(i-1)
                itemList.pop(i)
            else:
                i = i+1

        return itemList

    def _compile(self, expr:str):
        itemList = self.split(expr)
        itemList = self.addMulSymbol(itemList)

        for i in range(len(itemList)):
            item = itemList[i]
            if item.type == Compiler.ItemType.EXP:
                j = self._compile(item.string)
                if j[0] == Compiler.ItemType.EXP:
                    item.register = j[1]
                else:
                    itemList[i] = Compiler.Const(str(j[1]))

        itemList = self.calculate(itemList)

        if not len(itemList) == 1:
            raise Exception("Two separate expressions were entered.")
        if itemList[0].type == Compiler.ItemType.CST:
            return (Compiler.ItemType.CST, itemList[0].value)
        else:
            return (Compiler.ItemType.EXP, itemList[0].register)


    def compiler(self, expr:str):
        self.commandList = []
        self.maxRegisterNumber = 0
        self.registerList = [False for i in range(self.autoInitRegister)]

        i = self._compile(expr)
        if i[0] == Compiler.ItemType.EXP:
            self.o(Return(i[1]))
        else:
            self.o(ReturnNumber(i[1]))

        if self.maxRegisterNumber > self.autoInitRegister:
            i = InitRegister(self.maxRegisterNumber)
            self.commandList.insert(0, i)

        return self.commandList

